source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/MediaRenderer.java

```diff
@@ -0,0 +1,349 @@
+            m_renderedAdaptiveCard = renderedAdaptiveCard;
+            m_cardActionHandler = cardActionHandler;
+        }
+
+        @Override
+        public void onClick(View v)
+        {
+            if(m_allowedInlinePlayback)
+            {
+                if(((VideoViewResizeable) m_videoView).isAudioOnly())
+                {
+                    // if it's audio, just move the video to the top
+                    m_videoView.setZOrderOnTop(true);
+                }
+                else
+                {
+                    // If it's a video, hide the poster
+                    m_poster.setVisibility(View.GONE);
+                }
+
+                m_videoView.start();
+                m_videoView.setVisibility(View.VISIBLE);
+                m_playButon.setVisibility(View.GONE);
+            }
+
+            m_cardActionHandler.onMediaPlay(m_cardMediaElement, m_renderedAdaptiveCard);
+        }
+
+        private ImageView m_poster;
+        private ImageView m_playButon;
+        private VideoView m_videoView;
+        private boolean m_allowedInlinePlayback = true;
+
+        private BaseCardElement m_cardMediaElement;
+        private RenderedAdaptiveCard m_renderedAdaptiveCard;
+        private ICardActionHandler m_cardActionHandler;
+    }
+
+    private class MediaOnCompletionListener implements MediaPlayer.OnCompletionListener
+    {
+
+        MediaOnCompletionListener(BaseCardElement mediaElement, RenderedAdaptiveCard renderedAdaptiveCard, ICardActionHandler cardActionHandler)
+        {
+            m_cardMediaElement = mediaElement;
+            m_renderedAdaptiveCard = renderedAdaptiveCard;
+            m_cardActionHandler = cardActionHandler;
+        }
+
+        @Override
+        public void onCompletion(MediaPlayer mp)
+        {
+            m_cardActionHandler.onMediaStop(m_cardMediaElement, m_renderedAdaptiveCard);
+        }
+
+        private BaseCardElement m_cardMediaElement;
+        private RenderedAdaptiveCard m_renderedAdaptiveCard;
+        private ICardActionHandler m_cardActionHandler;
+    }
+
+    private class MediaOnPreparedListener implements MediaPlayer.OnPreparedListener
+    {
+        public MediaOnPreparedListener(MediaController mediaController)
+        {
+            m_mediaController = mediaController;
+        }
+
+        @Override
+        public void onPrepared(MediaPlayer mp)
+        {
+            m_mediaController.show(0);
+            m_mediaController.requestLayout();
+        }
+
+        private MediaController m_mediaController;
+    }
+
+    class OnlineFileCheckerTask extends AsyncTask<String, Void, Boolean>
+    {
+
+        protected Boolean doInBackground(String... url)
+        {
+            try
+            {
+                HttpURLConnection connection = (HttpURLConnection)new URL(url[0]).openConnection();
+                connection.setRequestMethod("HEAD");
+                return (connection.getResponseCode() == HttpURLConnection.HTTP_OK);
```

Comment:
what if we get a 301 redirect?

Reply:
not sure about it, Gilles or Andrew may be able to solve this question. There's no mention of it on the spec

---
In reply to: [202835408](https://github.com/Microsoft/AdaptiveCards/pull/1716#discussion_r202835408) [](ancestors = 202835408)

Reply:
@khouzam @andrewleader

Reply:
I would say that we should follow/support the redirect. I assume we support redirects today for images? Video ought to be the same. @almedina-ms any chance you can check the image redirect behavior to confirm? And thanks for catching this, I'll add it to the spec after you've confirmed whether images support redirects.

Reply:
in Android, the redirect in image is actually supported, but we check for Code 301 (HTTP_MOVED_PERM), Code 302 (HTTP_MOVED_TEMP) and Code 303 (HTTP_SEE_OTHER)

Reply:
Sounds good, let's do the same here for videos, I'll update the spec and let all the devs know that redirection needs to be supported

Reply:
Sounds good, let's do the same here for videos, I'll update the spec and let all the devs know that redirection needs to be supported